### ISSUE TITLE
[Bug?] Automatically processing the images after sync makes the image link unusable.

### ISSUE THREAD

**Issue by XiangFFF** [xRyul/obsidian-image-converter#123]
I am using remotely sync, and recently I find that after sync, image-converter will process the images updated again and make the image link unusable (not to mention I don't need the images to be compressed again)。I need the converter process only in the pasting the image behavior. Is there any way to fix this? THANKS A LOT!

**Comment by xRyul** [comment]
Can you provide some more details about your setup ? 

Do you sync to mobile or from one PC to another etc. do you have both instances of Obsidian open at the same time?

**Comment by XiangFFF** [comment]
The problem I describe happens when I sync between two PCs，both with Obsidian maintaining open. Additionally, I use other attachment management plugin to rename image while pasting, so my settings of image-converter are as follows:
1. Convert: webp, quality = 75
2. Output: default, keep original name
3. Settings: custom size = 200
4. Advanced: default

And it reminds me the rename behavior is done by other plugin, but is triggered by the converting behavior. So I disable that plugin, paste some new images, and try to sync again. Here is the new problem: converting behavior always happen when I have new images in the vault, and it rename the "111.image" to "111-1.image", so again, the image links "![[111.webp]]" fail. Can you fix it to stop the automatically converting behavior?

**Comment by xRyul** [comment]
You must wait for sync to complete on 1 device before opening it on another.

If you want to use Obsidian on both devices at the same time, then make sure that whatever sync method you use - it does not change note content and does not re-attach files to them. And make sure Image Converter settings are fully synced on both, otherwise you will have even more random results.

Taken all that into consideration, I still strongly advice against having 2 Obsidian instances open at the same time.

**Comment by XiangFFF** [comment]
OK，I guess I just have to disable the converter plugin before syncing, which should be the easiest way.

**Comment by xRyul** [comment]
Does "Remotely Save" have any status indicator?  

<img width="224" alt="image" src="https://github.com/user-attachments/assets/dac12523-1837-4a70-a7f1-b9cc8e93b3b1" />
  
For instance, Obsidian Sync has this green icon when it is done syncing, which indicates that everything got synced and should be safe to open Obsidian on another device

**Comment by XiangFFF** [comment]
Yes, Remotely Save has the similar status indicator, I just have to wait for the syncing process and then open the converter, the only problem is if I forget to disable, I have to open the attachment folder to find the altered pictures.

**Comment by BindingOx** [comment]
kept losing images and see there's no solution other than not having obsidian open on 2 computers. Hopefully we can get another solution.

**Comment by esodesod** [comment]
Hi! I'm experiencing something similar, using Obsidian Sync between two (2) macOS desktop versions of Obsidian (and iPhone/iOS and iPad/iPadOS, but no plugins enabled on iOS/iPadOS - only enabled on desktops)

**The "bug":** I have (some) orphaned images, which have the "old name" in the note, but the file has been renamed (filesystem), but the note does not reflect this change.

**Example:** I now have a note with the (invalid) embedded image link ![[20241218150519319-img.webp]], which was from the initial copy & paste (screenshot), but the file has been (incorrectly) renamed (again/afterwards), to `20241218174116733-img.webp`, which is now on my filesystem, but originally was `20241218150519319-img.webp`.

> [!NOTE]
> Based on the actual file metadata (macOS), I see that the timestamp for **created** (**15:05**) matches the original file name (20241218**1505**19319-img.webp) and that the **modified** timestamp (**17:41**) matches the new file name  (20241218**1741**16733-img.webp)

**Questions:**
- Do you think the above is due to a race condition in Obsidian Sync and/or Plugin re-processing the (webp) file, hence causing a new "rename" (without updating the note..)?
  - If the latter, why are webp files re-processed?
  - Can I simply add it to "skipPatterns"?
- Can we disable auto-processing somehow? (I don't see it in settings, but in data.json, I see `"autoRename": true`), if this will fully eliminate the situation(s) above; the goal is of course not to have many orphaned images lying around (and notes referencing the "non-existing old name.." :-).

I love this plugin, so thank you very much!

**Comment by xRyul** [comment]
@esodesod - the problem is mobile. And in particular that there is no `drag and drop` on mobile.  The options provided by obsidian e.g.: 'editor-drop' and 'editor-paste' work great on Desktop, but are not supported on mobile. Thus I have to resort to using `create` event handler before any of those events. `create` event handler enables image processing on mobile. But the problem is that it lacks context about how the file was created (e.g., pasted, dropped, or inserted via the paperclip) and if I run it directly before `editor-drop`, `editor-paste` events* - then it simply reacts to everything, including `sync`. *Hence why images get reprocessed**. Another problem is that I cannot simply remove `create` event from the code, as it is still needed for actual image processing. There are already quite a few safety checks in-place, which should prevent or at least minimize image processing - but clearly they are not enough. 

To anwser your question: 
- Yes, this is likely due to a race condition between Sync and the plugin's processing logic
- WebP files are processed because the create event handler doesn't discriminate between file types (it's needed for mobile support)
- While adding .webp to skipPatterns might help in some cases, it wouldn't address the underlying issue
- The `autoRename` setting controls renaming, but not the core image processing logic. You could technically disable all format conversions and set the quality to 100 to minimize processing, but the plugin would still react to `create` events. So you would need to disabled all conversion and renaming settings (convert to: same as original, quality 100, resize: none, file renaming: keep original, rename, use markdown links:disabled, non-destructive resize: none). Not to mention that some of these checks (and many others) are already being done automatically and applied in the background, in effort to minimize reprocessing. But timings might be off due to various factors, so `create` is still being slips through in certain cases. 

---

The solution could be to completely remove mobile support, and only allow Image Converter to work on Desktop. 

Or...

An option to allow image processing on mobile to only work on demand e.g. via right click - so then default Obsidian settings would handle image insertion into the note and if user would want to compress them on mobile they could right-click on the image and run appropriate command.

Basically, the core idea would be to minimize the number of events being constantly listened and only activate the necessary logic after user explicitly pastes or drops an image e.g.: register `create` event handler only AFTER a paste or drop event takes place.

On Desktop, in theory this would significantly reduce number of unnecessary event triggers and processing (in particular the problematic `create` event), and should lead to better overall plugin performance. 

On Mobile, because drop/paste is not supported via official Obsidian API, and we would not use problematic `create` event globally -  it would mean  images would not be processed immediately after adding via CMD+P or paperclip, instead user would now need to right click on the image and select conversion options there.

While this would be a compromise in terms of mobile functionality, it should provide much better reliability for image links across all platforms.

**Comment by esodesod** [comment]
Thank you for a **great explanation** of how the plugin automatically processes images, and highlighting the pros & cons of possible (future) changes; appreciate all the details!

As you mentioned, "on-demand processing" is exactly what I was wondering about, and maybe (personally) would prefer, especially if it could help to avoid possible orphaned images, e.g. images being renamed twice and note left with old image name (broken), as I experienced. That said, it goes both ways; some like it manual/controlled, and some like it automatic (I see both options as valuable, based on the use case).

To add a minor detail to my initial rename issue; guess you noticed based on my example, but I'm using `{date:YYYYMMDDHHmmssSSS}-img` as filename template. I initially thought this was based on `created` timestamp, but seeing the multiple renames (of the same file, twice), it seems to be based on the `modified` timestamp (even though I never "modified the image"). Point being; I thought the "replace existing" would simply add '-1, -2 -3 etc.' suffix to the end of the file, but results (when race condition or something occurs) is that the file is renamed (again) to the (latest) modified timestamp, which to my understanding, is due to the `create` event handler sees this as a new image (webp) file and then image converter processes the image file again, which is causing the image to be renamed twice. That said, I don't fully grasp why my note still referenced the "old image name" (the note should sync when the image is renamed, right?), but maybe not too important to dig into further, for now.

I've actually set up a testing vault with Obsidian Sync and Image Converter, trying to replicate/reproduce the "rename issue", now using version 1.2.22 (as I think my initial rename issues; and only a few - was from around my upgrade from 1.2.16 to 1.2.17, at least based on the sync log for manifest.json, for the plugin), but all my rename-, crop-, new image/paste-operations, etc. are working fine, is being synced fine, and working both ways (which is a good thing, I guess).

As suggested, I've now also disabled the "sync active community plugins" (selection using Obsidian Sync, even though I'd personally prefer to have it synced 1:1 between my desktop devices), and now disabled Image Converter on my "second desktop", leaving **Image Converter only enabled in one instance of Obsidian**, being my main desktop, which is usually always on (hence will get sync updates on-the-fly, if anything new comes in, from all devices). Hopefully, this workaround will not result in any orphaned images.

I'll keep an eye on the release notes and please let me know if I can assist in testing/verifying anything.

Again, thanks for your time! Happy new year (depending on TZ).

**Comment by xRyul** [comment]
For safety reasons, drop/paste/create listeners on mobile, Excalidraw and Canvas are now disabled. Conversion, compression, resizing is now only accessible via right-click or Obsidian file-menu. Which essentially fixes this issue.

**Comment by XiangFFF** [comment]
Awesome! Thanks for your effort!
